es_sample_shipment_timeliness function

philbam opened the pull request:
To Test:
1. Load required libraries: library(tidyverse), library(readr)
2. Load dataset as es_data
3. Dataset needs to contain the following columns: `collection.date`, `date.received.in.lab`, `es.lab.type`, `who.region`, `ADM0_NAME`
4. Test the function `es_sample_shipment_timeliness()`: usage pattern options include default all data (es_data) grouped by region and country; single or multiple regions; single or multiple countries; grouping by “region” or “country”; year filtering (specific period)
5. Check outputs:
- timeliness calculations and the median up to the current month per year.
- messages/info indicating missing dates or invalid date sequences

mcuadera commented on R/es_sample_shipment_timeliness.R:
```diff
@@ -0,0 +1,80 @@
+  else {years_to_use <- available_years}
+
+  es_data <- es_data |> dplyr::filter(lubridate::year(collection.date) %in% years_to_use)
+  if(nrow(es_data) == 0) {message("* No data available for the selected year(s) and location. Change filtering parameters or default to process all available data.")
```
No need to return a tibble

philbam replied:
Done

mcuadera commented on R/es_sample_shipment_timeliness.R:
```diff
@@ -0,0 +1,80 @@
+  if(total_samples != both_dates_available || invalid_records > 0 || has_no_lab_types) {
+    message(paste("* Data quality: Total ES Samples =", total_samples, "| Data with Collection and Lab receipt dates =", both_dates_available, "| Invalid Data (Lab receipt < Collection) =", invalid_records, "|| Valid Data Lab Type: Domestic =", domestic_count, "| International =", international_count, "| No Lab Type =", no_lab_type_count))}
+
+  # Determine grouping
```
group by region, country and month by default. No need to have a `group_by` parameter in the function

philbam replied:
Done

mcuadera commented on R/es_sample_shipment_timeliness.R:
```diff
@@ -0,0 +1,80 @@
+
+  # Calculate ES lab receipt timeliness results
+  result <- valid_data |>
+    dplyr::group_by(dplyr::across(dplyr::all_of(group_vars))) |>
```
just group by region, country name, and month

philbam replied:
Done

mcuadera commented on R/es_sample_shipment_timeliness.R:
```diff
@@ -0,0 +1,80 @@
+      pct_no_lab_type_within_7days = if(has_no_lab_types) ifelse(sum(lab_type == "no.es.lab.type") == 0, NA, round(sum(lab_type == "no.es.lab.type" & days_to_lab_receipt <= 7, na.rm = TRUE) / sum(lab_type == "no.es.lab.type") * 100, 1)) else NULL,
+      .groups = "keep")
+
+  # Calculate median up to current month for each year
```
We're not calculating up to the current month for this indicator, but rather this is a monthly indicator. Here's the note on this indicator:

> Monthly 
> Compare with average/median of same month for the year before.

You could calculate the monthly up to the end date still. That's fine

philbam replied:
Done, now calculates monthly up to the end date's month.

philbam commented on R/es_sample_shipment_timeliness.R:
```diff
@@ -0,0 +1,80 @@
+#' #' Timeliness of ES Shipment
+#' @description Days between ES sample collection and lab receipt of shipped samples (Target: 3 days for Domestic Labs and 7 days for International Labs)
+#' @param es_data,region,country,year,group_by,end_date Data frame and optional filters: region(s), country(ies), year(s) (single year for starting year or range), grouping (group by = ""), and reference date (default Sys.Date())
```
Updated @param tags, simplified parameters structure by removing region, country, year, and group_by, default grouping now set to region, country, and month, and added \dontrun{} examples.

philbam commented on R/es_sample_shipment_timeliness.R:
```diff
@@ -0,0 +1,80 @@
+  if(!is.null(region)) es_data <- es_data |> dplyr::filter(toupper(trimws(who.region)) %in% toupper(trimws(region)))
+  if(!is.null(country)) es_data <- es_data |> dplyr::filter(toupper(trimws(ADM0_NAME)) %in% toupper(trimws(country)))
+
+  # Determine years and filter
```
Modified to allow only end date.

philbam commented on R/es_sample_shipment_timeliness.R:
```diff
@@ -0,0 +1,80 @@
+  if(nrow(es_data) == 0) {message("* No data available for the selected year(s) and location. Change filtering parameters or default to process all available data.")
+    return(dplyr::tibble())}
+
+  # Data validation and lab type processing
```
Created separate activity_dates_data_validation() function and integrated lab_loc join at the beginning to establish es.lab.type column before validation.

philbam commented:
Intend to close this PR as I've moved the refactored version with updates addressing your review feedback to the branch that aligns with the issue naming. Please continue the review in PR #26 which includes the requested changes: simplified parameter structure, default grouping by region, country, and month, integration with the new separate validation function, corrected end_date filtering logic, and improved documentation formatting. Thank you for the detailed feedback.

philbam commented:
Closing this PR: incorporated into a single ES timeliness function, refer to PR #27
